controller/backing_image_controller.go

```diff
@@ -279,6 +287,32 @@ func (bic *BackingImageController) syncBackingImage(key string) (err error) {
 	return nil
 }
 
+func (bic *BackingImageController) IsBackingImageDataSourceCleanup(bi *longhorn.BackingImage) (finished bool, err error) {
+	bids, err := bic.ds.GetBackingImageDataSource(bi.Name)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return false, err
+	}
+
+	if bids.Spec.FileTransferred {
+		return true, nil
+	}
+
+	// when bids status is ready or ready for transfer, there is already no tmp file left
+	if bids.Status.CurrentState == longhorn.BackingImageStateFailedAndCleanUp ||
+		bids.Status.CurrentState == longhorn.BackingImageStateReady ||
+		bids.Status.CurrentState == longhorn.BackingImageStateReadyForTransfer {
+		return true, nil
+	}
+
+	// mark the status to failed so manager can clean up the orphan tmp file
+	bids.Status.Message = "backing image is deleted, try to cleaning up the tmp file of data source"
+	bids.Status.CurrentState = longhorn.BackingImageStateFailed
+	if _, err = bic.ds.UpdateBackingImageDataSourceStatus(bids); err != nil {
+		return false, err
+	}
+	return false, nil
```

Comment:
It's unusual to return false w/o error if the function/method support returning bool and error together. However, here it's quite special usage, because in fact there are two conditions here instead one, it would cause some confusion initially.

- backing image has been cleaned up 
- backing image was not cleaned up but has been requested to clean up 
  - but having error while updating the data source status
  - updated the status successfully

Ideally there should be three returned variables to make things clean. (`cleanUp`, `requested`, `error`). However, it's fine for this PR.

Reply:
I think you misunderstood my comment. It's more about general coding patterns, when returning a false variable, it should have an error variable to explain why it's failed. This was why I said should be better to clarify the returned variables if could. 

Doing asynchronous fix is totally correct here.

Reply:
yeah, I get your point now,

Maybe a better way is to remove the return err variable to make this function only return true or false,
if there is any error, we just log error and return false to wait next reconcile loop

Thanks for the advise

Reply:
Yeah this is another way but as a dependent function/method it's better to return error to let callers decide what to do.

Again, this is minor stuff but a basic guide to share with you 🙂